pkg/util/winutil/service.go

```diff
@@ -92,150 +102,156 @@ func StartService(serviceName string, serviceArgs ...string) error {
-//	5. Core agent now starts process-agent dependent service
-//	6a. Older implementation will stop trace-agent dependent service
-//	6b. Older implementation will stop trace-agent and process-agent dependent service
-//	7a. Older implementation will stop core agent service but it will fail because
-//	    process-agent dependent service is still running
-//	7b. New implementation will stop core agent service and it will succeed since
-//	    no dependent services are running.
-//
-// This will not solve 100% of the edge cases, but it is still a good and simple improvement,
-// which will eliminate ebove mentioned edge case and a few more. Full race condition fixes
-// will be implemented in the next iterations.
+//		For illustration purposes here is an example of one specific race condition:
+//		1. The core agent (datadogagent) service is starting
+//		2. Starting core agent starts trace-agent dependent service
+//		3. Agent's restart-service CLI command started
+//		4. Agent get list all dependent services (e.g., trace-agent, process-agent, system-probe
+//		5. Agent's restart-service CLI command enumerate all dependent services and if they are
+//	    running it will stop them.
+//		6. Just before restart-service CLI command will stop core agent service, the service
+//	    is still starting and it may start trace-agent dependent service.
+//		7. Attempt to stop core agent service will fail because a dependent service is "still" running
 //
 // Callback is invoked to help unit tests to setup race condition in deterministic way
-func stopServiceInternal(serviceName string, enumState enumServiceState, callback stopServiceCallback) error {
-	deps, err := ListDependentServices(serviceName, enumState)
+func doStopServiceWithDependencies(manager *mgr.Mgr, service *mgr.Service, serviceName string,
+	depenStatus svc.ActivityStatus, callback stopServiceCallback) error {
+
+	ctx, cancel := context.WithTimeout(context.Background(), defaultServiceCommandTimeout*time.Second)
+	defer cancel()
+
+	// open dependent services
+	depServiceNames, err := service.ListDependentServices(depenStatus)
 	if err != nil {
 		return fmt.Errorf("could not list dependent services for %s: %v", serviceName, err)
 	}
 	if callback != nil {
 		callback.afterDependentsEnumeration()
 	}
-
-	for _, dep := range deps {
-		if callback != nil {
-			callback.beforeStopService(dep.serviceName)
-		}
-		err = doStopService(dep.serviceName)
+	var depServices []*mgr.Service
+	for _, depServiceName := range depServiceNames {
+		depService, err := OpenService(manager, depServiceName, windows.SERVICE_STOP|windows.SERVICE_QUERY_STATUS)
 		if err != nil {
-			return fmt.Errorf("could not stop service %s: %v", dep.serviceName, err)
+			return fmt.Errorf("could open service %s: %w", depServiceName, err)
 		}
+		depServices = append(depServices, depService)
+		defer depService.Close()
 	}
 
-	if callback != nil {
-		callback.beforeStopService(serviceName)
-	}
+	for {
```

Comment:
<!-- Questions are appropriate if you have a potential concern but are not quite sure if it's relevant or not. Asking the author for clarification or investigation can lead to a quick resolution. -->
**❓ question**
have you considered using out of the box retry libraries like `backoff.Retry`? Wonder if we should sleep between each retry to reduce CPU impact

Reply:
I do not think so in this case. First it is capped by timeout, currently 30 seconds - it will be short and anyway and second ability for restart to be "tight" is relatively important, Second, doStopService I think waits 300 ms on each individual service stop, external, meaning this loop is to "detect" race condition which happens and immediately try to address it.